### ISSUE TITLE
approx_size() in mpmc_bounded_q.h

### ISSUE THREAD

**Issue by Subenle** [gabime/spdlog#546]
In file mpmc_bounded_q.h:
```
std::atomic<size_t>     enqueue_pos_;
std::atomic<size_t>     dequeue_pos_;
```
Every time after enqueue/dequeue a message, `enqueue_pos_ / dequeue_pos_` will increse by 1 .

When `enqueue_pos_` exceed the limits of `size_t`, it will be 0 again. However, `dequeue_pos_` may not exceed the limits,  function `approx_size()` will also return 0 in this condition.

The function `approx_size()` is here: https://github.com/gabime/spdlog/blob/master/include/spdlog/details/mpmc_bounded_q.h#L136

So, in file `async_log_helper.h` , function `wait_empty_q` may also be not true.

`wait_empty_q`:
https://github.com/gabime/spdlog/blob/master/include/spdlog/details/async_log_helper.h#L384

**Comment by gabime** [comment]
I am not sure how enqueue_pos_ can exceed size_t - the queue size is limited by max size(also of type size_t).

**Comment by Subenle** [comment]
@gabime 
Let me show you:  
`enqueue_pos_ / enqueue_pos_` are init as `0`.
Every time the data is enqueued in `pos` successfully, then `enqueue_pos_ = pos + 1`, see the code bellow:
```
enqueue_pos_.compare_exchange_weak(pos, pos + 1, std::memory_order_relaxed)
```
`enqueue_pos_` has not been limited to be smaller than `max_size_`.  

We get `cell = &buffer_[pos & buffer_mask_];`, here `pos & buffer_mask_` will always be in [0, max_size).

We guess the max queue size is `max_size_ = 4`, after enqueue 4 data successfully, the `enqueue_pos_ = 4`, then, enqueue a new data, we get: 
```
pos = enqueue_pos_.load(std::memory_order_relaxed);   // pos = 4
cell = &buffer_[pos & buffer_mask_];  // cell = &buffer_[4 & 3] = &buffer_[0]
```
the new data will be put in buffer_[0].  Then `enqueue_pos_ = 5`, and next time the data will be enqueue to buffer_[5 & 3] = buffer_[2].

It means that pos can be bigger than max_size, but `pos & buffer_mask_` will be limited during [1, max_size)

**Comment by gabime** [comment]
Right. 
So a possible fix would be:

```c++
if(last_pos < first_pos)
   return (size_t - first_pos + last_pos);
```

But we need to be careful here, I think  last_pos can be temporarly bigger than first_pos because of the use of the std::memory_order_relaxed model. 
Here is the [converstation](https://groups.google.com/forum/m/#!topic/lock-free/XPEAkJwFsKw) I had with the inventor of this lockfree queue regarding this issue.

**Comment by gabime** [comment]
Fixed in f5fe681a4193866cda6fd8deb1c7026e4491d75d
